TESTS/netsocket/udp/udpsocket_echotest.cpp

```diff
@@ -63,37 +63,45 @@ void UDPSOCKET_ECHOTEST()
     NetworkInterface::get_default_instance()->gethostbyname(ECHO_SERVER_ADDR, &udp_addr);
     udp_addr.set_port(ECHO_SERVER_PORT);
 
-    UDPSocket sock;
-    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, sock.open(NetworkInterface::get_default_instance()));
+    UDPSocket socket;
+    TEST_ASSERT_EQUAL(NSAPI_ERROR_OK, socket.open(NetworkInterface::get_default_instance()));
+    socket.set_timeout(WAIT2RECV_TIMEOUT);
 
     int recvd;
     int sent;
     int packets_sent = 0;
     int packets_recv = 0;
     for (int s_idx = 0; s_idx < sizeof(pkt_sizes) / sizeof(*pkt_sizes); ++s_idx) {
+        recvd = 0;
         int pkt_s = pkt_sizes[s_idx];
 
         fill_tx_buffer_ascii(tx_buffer, BUFF_SIZE);
         int packets_sent_prev = packets_sent;
 
-        for (int retry_cnt = 0; retry_cnt <= 2; retry_cnt++) {
+        for (int retry_cnt = 0; retry_cnt <= RETRIES; retry_cnt++) {
             memset(rx_buffer, 0, BUFF_SIZE);
-            sent = sock.sendto(udp_addr, tx_buffer, pkt_s);
+            sent = socket.sendto(udp_addr, tx_buffer, pkt_s);
             if (check_oversized_packets(sent, pkt_s)) {
                 TEST_IGNORE_MESSAGE("This device does not handle oversized packets");
             } else if (sent == pkt_s) {
                 packets_sent++;
+                break;
             } else {
                 printf("[Round#%02d - Sender] error, returned %d\n", s_idx, sent);
                 continue;
             }
-            recvd = sock.recvfrom(NULL, rx_buffer, pkt_s);
-            if (recvd == pkt_s) {
+        }
+
+        for (int retry_cnt = 0; retry_cnt <= RETRIES; retry_cnt++) {
+            recvd += socket.recvfrom(NULL, rx_buffer + recvd, pkt_s - recvd);
```

Comment:
No.
This cannot be accepted.

UDP payload is a datagram, and by specification, it should maintain the datagram boundaries.
We should NOT append any received data with the previous packet.

Reply:
I was mislead by the [UDPSocket::recv](https://github.com/ARMmbed/mbed-os/blob/e7bc177b20427197763e1cd47261b04bf18db64e/features/netsocket/UDPSocket.h#L146) doxygen which says 

> By default, recv blocks until some data is received.

GR_LYCHEE slices the datagrams. It uses [ESP32 driver](https://github.com/d-kato/esp32-driver). The driver itself does not limit the size of the datagram, neither does it implement any slicing mechanism internally. It is coded similarly to the ESP8266 driver, but I noticed it doesn't differentiate between TCP and UDP, while [ESP8266 does](https://github.com/ARMmbed/mbed-os/blob/e7bc177b20427197763e1cd47261b04bf18db64e/components/wifi/esp8266-driver/ESP8266/ESP8266.cpp#L845).
Perhaps the part that put a sliced datagram together is missing.
I will post an issue in the ESP32 driver, asking for a fix and pointing to the ESP8266 as a reference.

Reply:
In theory, even `Socket::recv()` should give you the whole datagram.
The Socket API is identical between TCP and UDP, so you can use `UDPSocket::connect()` and then just `::send()` and ``recv()` and it works similarly as `recvfrom()`.
Only difference is basically how the peer address is defined.

So overall, receiving or sending UDP should not do any slicing as it is a datagram protocol and therefore should respect the boundaries.
This does not apply to TCP which is a stream, and therefore does not enforce any boundaries.